<!-- FroboScout launch file -->

<launch>
	<!-- FroboScout parameters --> 
	<param name="diff_steer_wheel_distance" value="0.5"/>
	<param name="ticks_per_meter_left" value="2627" />
	<param name="ticks_per_meter_right" value="2627" />

	<!-- node pkg="rosbag" name="rosbag_record" type="record" args="-a"/ -->

	<group ns="fmSensors">
	</group>

	<group ns="fmProcessors">
		<!-- calculate odometry based on encoders and IMU yaw axis orientation -->
		<node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" >
			<param name="enc_left_sub" value="/fmInformation/encoder_left" />
			<param name="enc_right_sub" value="/fmInformation/encoder_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_pub" value="/fmKnowledge/pose" />
			<param name="odom_estimate_frame" value="/odom_combined" />
			<param name="vehicle_frame" value="base_footprint" />
			<param name="encoder_output" value="absolute" />
			<param name="max_ticks_per_update" value="200" />
			<param name="yaw_angle_source" value="imu_angular_velocity" />
			<param name="imu_angular_velocity_yaw_axis" value="y" />
		</node>
	</group>

	<group ns="fmDecisionMakers">
	</group>

	<group ns="fmExecutors">
		<node pkg="sdu_surveying" name="velocity_set" type="cmd_vel_publish.py" output="screen" >
			<param name="deadman_sub" value="/fmCommand/deadman"/>
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel"/>
			<param name="update_interval" value="0.1"/>
			<param name="linear_velocity" value="0.25"/>
			<param name="angular_velocity" value="0.0"/>
		</node>
	</group>

	<group ns="fmControllers">
	</group>

	<group ns="fmLib">
		<!-- FroboScout interface node -->
		<node pkg="cetus_froboscout" name="froboscout" type="froboscout_interface_node.py" output="screen" >
    		<param name="deadman_sub" value="/fmCommand/deadman" />
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="enc_left_pub" value="/fmInformation/encoder_left" />
			<param name="enc_right_pub" value="/fmInformation/encoder_right" />
			<param name="wheel_status_left_pub" value="/fmInformation/wheel_status_left" />
			<param name="wheel_status_right_pub" value="/fmInformation/wheel_status_right" />
			<param name="wheel_feedback_left_pub" value="/fmInformation/wheel_feedback_left" />
			<param name="wheel_feedback_right_pub" value="/fmInformation/wheel_feedback_right" />
			<param name="wheel_left_sub" value="/fmData/wheel_left_nmea_in" />	 
			<param name="wheel_left_pub" value="/fmSignal/wheel_left_nmea_out" /> 
			<param name="wheel_right_sub" value="/fmData/wheel_right_nmea_in" />	 
			<param name="wheel_right_pub" value="/fmSignal/wheel_right_nmea_out" /> 
			<param name="publish_wheel_status_rate" value="5" />
			<param name="publish_wheel_feedback_rate" value="50" />
            <param name="max_linear_acceleration" value="3.0"/>
            <param name="max_angular_acceleration" value="6.0"/>
			<param name="wheel_left_kp" value="9" /> 
			<param name="wheel_left_ki" value="15" /> 
			<param name="wheel_left_kd" value="2" /> 
			<param name="wheel_left_max_integral_output" value="50.0" /> 
			<param name="wheel_right_kp" value="9" /> 
			<param name="wheel_right_ki" value="15" /> 
			<param name="wheel_right_kd" value="2" /> 
			<param name="wheel_right_max_integral_output" value="50.0" /> 
		</node>

		<!-- nmea protocol handler for the FroboScout left wheel module -->
		<node pkg="serial_nmea" name="left_wheel_nmea" type="nmea_node" output="screen" respawn="true">
			<param name="str_to_msg_sub" value="/fmData/wheel_left_rx" />	 
			<param name="str_to_msg_pub" value="/fmData/wheel_left_nmea_in" />	 
			<param name="msg_to_str_sub" value="/fmSignal/wheel_left_nmea_out" /> 
			<param name="msg_to_str_pub" value="/fmData/wheel_left_tx" />	 
		</node> 

		<!-- nmea protocol handler for the FroboScout right wheel module -->
		<node pkg="serial_nmea" name="right_wheel_nmea" type="nmea_node" output="screen" respawn="true">
			<param name="str_to_msg_sub" value="/fmData/wheel_right_rx" />	 
			<param name="str_to_msg_pub" value="/fmData/wheel_right_nmea_in" />	 
			<param name="msg_to_str_sub" value="/fmSignal/wheel_right_nmea_out" /> 
			<param name="msg_to_str_pub" value="/fmData/wheel_right_tx" />	 
		</node> 

		<!-- serial communication with the FroboScout left wheel module -->
		<node pkg="serial_string" name="left_wheel_ser" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/wheel_left_rx" />	 
			<param name="subscriber_topic" value="/fmData/wheel_left_tx" />
			<param name="device" value="/dev/wheel_left" />	
			<param name="baudrate" value="57600" />
			<param name="termination_character" value="10"/>
		</node> 

		<!-- serial communication with the FroboScout right wheel module -->
		<node pkg="serial_string" name="right_wheel_ser" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/wheel_right_rx" />	 
			<param name="subscriber_topic" value="/fmData/wheel_right_tx" />
			<param name="device" value="/dev/wheel_right" />	
			<param name="baudrate" value="57600" />
			<param name="termination_character" value="10"/>
		</node> 
	</group>
</launch>

